src/index/sparse/sparse_inverted_index.h

```diff
@@ -0,0 +1,430 @@
+                }
+            }
+        }
+
+        n_rows_ += rows;
+        nnz_ += nnz;
+        return Status::success;
+    }
+
+    void
+    Search(const void* query_csr_matrix, int64_t q_id, size_t k, float drop_ratio_search, float* distances,
+           label_t* labels, size_t refine_factor, const BitsetView& bitset) const {
+        size_t len;
+        const IndicesT* indices;
+        const T* data;
+        get_row<T>(query_csr_matrix, q_id, len, indices, data);
+
+        // initially set result distances to NaN and labels to -1
+        std::fill(distances, distances + k, std::numeric_limits<float>::quiet_NaN());
+        std::fill(labels, labels + k, -1);
+        if (len == 0) {
+            return;
+        }
+
+        // if no data was dropped during both build and search, no refinement is needed.
+        if (drop_ratio_build_ == 0 && drop_ratio_search == 0) {
+            refine_factor = 1;
+        }
+
+        std::vector<std::pair<IndicesT, T>> q_vec(len);
+        for (size_t i = 0; i < len; ++i) {
+            q_vec[i] = std::make_pair(indices[i], data[i]);
+        }
+        std::sort(q_vec.begin(), q_vec.end(),
+                  [](const auto& lhs, const auto& rhs) { return std::abs(lhs.second) > std::abs(rhs.second); });
+        while (!q_vec.empty() && q_vec[0].second * drop_ratio_search > q_vec.back().second) {
+            q_vec.pop_back();
+        }
+
+        MaxMinHeap<T> heap(k * refine_factor);
+        if (!use_wand_) {
+            search_brute_force(q_vec, heap, bitset);
+        } else {
+            search_wand(q_vec, heap, bitset);
+        }
+
+        // no refinement needed
+        if (refine_factor == 1) {
+            collect_result(heap, distances, labels);
+        } else {
+            // TODO tweak the map buckets number for best performance
+            std::unordered_map<IndicesT, T> q_map(4 * len);
+            for (size_t i = 0; i < len; ++i) {
+                q_map[indices[i]] = data[i];
+            }
+            refine_and_collect(q_map, heap, k, distances, labels);
+        }
+    }
+
+    [[nodiscard]] size_t
+    size() const {
+        size_t res = 0;
+        res += sizeof(*this);
+        res += sizeof(T) * data_.capacity();
+        res += sizeof(IndicesT) * indices_.capacity();
+        res += sizeof(IndPtrT) * indptr_.capacity();
+        res += sizeof(std::vector<Neighbor<T>>) * inverted_lut_.capacity();
+        for (auto& lut : inverted_lut_) {
+            res += sizeof(Neighbor<T>) * lut.capacity();
+        }
+        if (use_wand_) {
+            res += sizeof(T) * max_in_dim_.capacity();
+        }
+        return res;
+    }
+
+    [[nodiscard]] size_t
+    n_rows() const {
+        return n_rows_;
+    }
+
+    [[nodiscard]] size_t
+    n_cols() const {
+        return n_cols_;
+    }
+
+ private:
+    [[nodiscard]] float
+    dot_product(const std::unordered_map<IndicesT, T>& q_map, table_t u) const {
```

Comment:
std::unordered_map seems with a bad performance as I tested. It's better to use two arrays `indices` and `data`